### ISSUE TITLE
AGS 4: in new compiler (optionally) disable imports of non-pointer managed objects

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1729]
CC: @fernewelten 

Reading the [description of a new compiler's syntax extensions](https://github.com/adventuregamestudio/ags/wiki/Features-of-the-new-Script-Compiler) reminded me of one problem.

> The big exception is import declarations. *In import declarations, it is legal to declare managed structs that aren't pointers*, and automatically generated AGS code will actually do that, so the compiler does not imply any * for things that are declared import. Any pointers must be explicitly written out in import declarations.

If I remember correctly, this is related to the several global arrays and objects in script API:
```
import Object object[AGS_MAX_OBJECTS];
import Hotspot hotspot[AGS_MAX_HOTSPOTS];
import Region region[AGS_MAX_REGIONS];
```
and so on - in [generated arrays for characters]() etc:
https://github.com/adventuregamestudio/ags/blob/1bb0320be16d3b98ef2a8f832ebe10f39c7fcf6a/Editor/AGS.Editor/Tasks.cs#L719-L724
https://github.com/adventuregamestudio/ags/blob/1bb0320be16d3b98ef2a8f832ebe10f39c7fcf6a/Editor/AGS.Editor/Tasks.cs#L739-L742

---

This kind of declaration goes against the principle of a managed object, which is only supposed to be passed and addressed by a pointer. The reason this is done so, iirc, is historical: it's likely that *originally these structs were not managed but regular structs*, and as AGS script did never support raw pointers, these objects were presented as if they were allocated on global script memory (in reality they were part of the internal engine memory).

Such way to declare objects in API causes a serious compatibility problem: as script calculates element offsets based on their sizes, thus changing a size of such struct would break any previously compiled scripts, making it impossible to run older games.

I believe that for AGS 4.* branch we should _completely get rid of this kind of API_, replacing struct arrays with either pointer-to-struct arrays, or indexed attributes again returning pointers.

Changing API is one task; but along with that I think that the new compiler must enforce a proper syntax which would forbid such constructs. Or, in other words, in accordance to the new syntax rule of "Auto pointering", read the `import ManagedType var;` and `import ManagedType func();` as pointers and functions returning pointers correspondingly. Thus making "auto pointering" fully consistent.

Since the new compiler may still be used to test older scripts, perhaps this may be done using a new compilation option: allow or disallow importing managed objects without a pointer. In AGS 4 we will keep this at "allow" until the API is properly modified.

**Comment by ivan-mogilko** [comment]
Also, in the existing engine API the only case when such syntax is used is when declaring global objects or arrays. But this is not permitted for functions return values, or arguments, nor attributes, as ags script does not allow to pass a managed struct by value, only access one.

I think that, if it's possible to do without much hassle, perhaps the compiler could only apply this distinction between having a pointer to or an object itself under following conditions:
1. This has to be a `builtin` struct;
2. This has to be a variable or array declaration.

In all other cases it should consider expression `ManagedType` an "autopointer".

**Comment by fernewelten** [comment]
Good thoughts!

I can do that "Only builtin, only variables or arrays" restriction, and I can also do that compiler option.

**Comment by ericoporto** [comment]
I think we managed to adjust all of those in ags4 to use pointers so the only thing now is the compiler behavior.

**Comment by fernewelten** [comment]
> I think we managed to adjust all of those in ags4 to use pointers so the only thing now is the compiler behavior.

OK, will do.

**Comment by ivan-mogilko** [comment]
Implemented by #2513
